can-view-scope.js

```diff
@@ -34,30 +44,155 @@ function Scope(context, parent, meta) {
+			info.isContextBased = true;
+			if(parentContext !== undefined) {
+				info.parentContextWalkCount++;
+			}
+			return "";
+		});
+		// ../..
+		if(info.remainingKey === "..") {
+			info.parentContextWalkCount++;
+			info.remainingKey = "this";
+		}
+		else if(info.remainingKey === "." || info.remainingKey === "") {
+			info.remainingKey = "this";
+		}
+
+		if(info.remainingKey === "this") {
+			info.isContextBased = true;
+		}
 		return info;
+	},
+	// converts
+	// 	../foo -> ./foo
+	//  ../. -> .
+	//  .. -> .
+	//  .././foo -> ./foo
+	// ../../foo -> ../foo
+	removeLeadingParentWalk: function(key, keyInfo){
+		if(keyInfo.isParentContext) {
+			return ".";
+		}
+		var remaining = key.substr(3);
+		if(remaining === "" || remaining === ".") {
+			return ".";
+		}
+		var after = remaining.substr(0,2);
+		if(after === "..") {
+			return remaining;
+		}
+		else if(after !== "./") {
+			return "./"+remaining;
+		} else {
+			return remaining;
+		}
+
+	},
+	isTemplateContextOrCanNotHaveProperties: function(currentScope){
+		var currentContext = currentScope._context;
+		if(currentContext instanceof TemplateContext) {
+			return true;
+		} else if( !canHaveProperties(currentContext) ) {
+			return true;
+		}
+		return false;
+	},
+	shouldSkipIfSpecial: function(currentScope){
+		var isSpecialContext = currentScope._meta.special === true;
+		if (isSpecialContext === true) {
+			return true;
+		}
+		if( Scope.isTemplateContextOrCanNotHaveProperties(currentScope) ) {
+			return true;
+		}
+		return false;
+	},
+	shouldSkipEverythingButSpecial: function(currentScope){
+		var isSpecialContext = currentScope._meta.special === true;
+		if (isSpecialContext === false) {
+			return true;
+		}
+		if( Scope.isTemplateContextOrCanNotHaveProperties(currentScope) ) {
+			return true;
+		}
+		return false;
+	},
+	// TODO: Remove?
+	// This will keep checking until we hit the next normal context
+	makeShouldExitOnSecondNormalContext: function(){
+		var foundNormalContext = false;
+		return function shouldExitOnSecondNormalContext(currentScope){
+			var isNormalContext = !currentScope.isSpecial();
+			var shouldExit = isNormalContext && foundNormalContext;
+			// leaks some state
+			if(isNormalContext) {
+				foundNormalContext = true;
+			}
+			return shouldExit;
+		};
+	},
+	// This will not check anything after the first normal context
+	makeShouldExitAfterFirstNormalContext: function(){
+		var foundNormalContext = false;
+		return function shouldExitAfterFirstNormalContext(currentScope){
+			if(foundNormalContext) {
+				return true;
+			}
+			var isNormalContext = !currentScope.isSpecial();
+			// leaks some state
+			if(isNormalContext) {
+				foundNormalContext = true;
+			}
+			return false;
+		};
+	},
+	makeShouldSkipSpecialContexts: function(parentContextWalkCount){
+		var walkCount = parentContextWalkCount || 0;
+		return function(currentScope){
+			if(currentScope.isSpecial()) {
+				return true;
+			}
+			walkCount--;
```

Comment:
Can you explain why this is decrementing the `walkCount`? I'm not sure exactly what's going on here.

Reply:
This is used for something like `../../../foo`, which would result in a `parentContextWalkCount` of 3, one for each `../`.  

It uses that number as a counter to know when it should start searching for `foo`.